src/protocol/receiver/mod.rs

```diff
@@ -258,12 +257,9 @@ impl Receiver {
         if self.timers.nak.check_expired(now).is_some() {
             self.on_nak_event(now);
         }
-        if self.timers.exp.check_expired(now).is_some() {
-            // TODO: what about EXP?
-            self.on_exp_event(now);
-        }
-        if self.timers.timeout.check_expired(now).is_some() {
-            self.on_timeout();
+        if self.timers.exp.check_expired(now).is_some() && self.on_exp_event(now) {
+            self.shutdown_flag = true;
+            info!("16 consequtive exp events, timeout");
```

Comment:
This does look correct, according to the spec. Also, just make on_exp_event conditionally set the shutdown_flag. No need to have it return bool and complicate the semantics, none of the other “event” functions have a return value

Reply:
Good point. Will do.